### ISSUE TITLE
RFE: Add "rename" subcommand for better Docker compatibility

### ISSUE THREAD

**Issue by cjeanner** [containers/podman#1925]
# **kind feature**

**Description**

Docker has a "rename" subcommand allowing to change a container name. This feature is missing in podman, and is useful in some cases.

Thank you!

**Comment by EmilienM** [comment]
I agree the feature from [docker](https://docs.docker.com/engine/reference/commandline/rename/) would be useful. Shutting down and removing the container then creating a new one isn't a valid answer when you run containers in production.

**Comment by mheon** [comment]
I was thinking about a `podman alias` command that would let you add/remove aliases that would act as extra names for a container. Changing a container's actual name would be very difficult for us do to the internal structure of containers, but I can easily let you add/remove additional names from existing containers.

**Comment by cjeanner** [comment]
@mheon hello :).
Sooo. an alias might work, but we'll need to find a way to make this alias work for tasks like "podman start <name>" and the like.
Also, it would be needed to set a "preferred alias" or something that shows up in a `podman ps` output as the container name, else we will have "some" issues finding what container is running and so on.

Fact is, we really rely on the container name for a lot of tasks, and we need it to be as stable as possible - like docker ;).
Usages include:
- podman ps --format "{{ .Names }}"
- podman exec "..." NAME
- podman start "..." NAME
- podman stop "..." NAME
and so on. So the name is important, and as we have a thing that manages both docker and podman engines, we'd like to get a really good compatibility in order to avoid code duplication, weird "if/then/else" and things like that that make the code hard to follow/maintain.

I don't know if @EmilienM has other thoughts.. ?

**Comment by baude** [comment]
During a rename, can the container ID change ?

**Comment by cjeanner** [comment]
@baude I think the id can change - at least, I see no reason for us to base our workflows on ids as they aren't predictable, on the contrary to the name (hence the need of the "rename" :) ).

**Comment by mheon** [comment]
Aliases would work like container names for everything except PS, where
we'd probably need a separate column for them (would be good enough for
grep, but probably break things if you're parsing the output of PS). Of
course, if you wanted to filter to a single container based on name/alias,
we could make that work exactly like it does now

**Comment by cjeanner** [comment]
@mheon hmm ok. well, I think we can give a try?

**Comment by cjeanner** [comment]
Hello @mheon - any news for this? :)

Cheers,

C.

**Comment by cjeanner** [comment]
hello @mheon - any news? We have a really ugly work-around, and would really appreciate something more "nice" :).

**Comment by mheon** [comment]
I'll bump it up the to-do list again, but no promises - things consistently keep managing to explode and require attention.

**Comment by rhatdan** [comment]
@mheon This probably goes along with the clone stuff.

**Comment by rhatdan** [comment]
Seems like this would be something for a new developer to take a stab at?

**Comment by mheon** [comment]
Agree - if we have any interns coming on, or anyone who wants to take a
first issue, a clone command is a good starting point.

**Comment by TomSweeneyRedHat** [comment]
@dougsland interested in trying a Podman issue?  Then we can get you a Podman sticker too!

**Comment by dougsland** [comment]
@TomSweeneyRedHat sure thing, let me check and dig this.

**Comment by weirdwiz** [comment]
@dougsland are you working on this issue? can I take this up?

**Comment by dougsland** [comment]
> @dougsland are you working on this issue? can I take this up?

@weirdwiz I am in a conference right now but my plan is continue on this next week. However, if you have any patch ready to go fell free to submit.

**Comment by weirdwiz** [comment]
@dougsland I see, I will leave you to it then as you are already working on it :grinning:

**Comment by rhatdan** [comment]
@dougsland @ashley-cui @weirdwiz @QiWang19 Since this never made progress could someone verify that they are working on it, or grab it for themselves.

**Comment by QiWang19** [comment]
I can take this if @dougsland  hasn't stated.

**Comment by mheon** [comment]
To be 100% clear here: I don't think `podman rename` makes sense in current context. I would greatly prefer a `podman container clone` that makes a copy of an existing container, with changes.

**Comment by mheon** [comment]
(Aliases are also a potential option, but I think we should start with `clone`)

**Comment by rhatdan** [comment]
I am fine with this, as long as we have an alternative.

**Comment by TomSweeneyRedHat** [comment]
FWIW, Docker has a rename....  https://docs.docker.com/engine/reference/commandline/rename/

**Comment by vrothberg** [comment]
@mheon @rhatdan, let's find a consensus on how to proceed. Instead of `rename` we would add a `clone` that might expose an optional `--name` flag?

@QiWang19, do you have enough cycles to work on that or shall we add a community tag?

**Comment by mheon** [comment]
@vrothberg I'm poking at clone in spare cycles, but it's a bit of a monster - needs a pretty significant rewrite of our parsing code to handle it.

**Comment by rhatdan** [comment]
Can we start small, and only allow clone to change the name in the beginning.  Then we can add options as people ask.

Not sure we need full support for every option.

**Comment by mheon** [comment]
Sure, I can get something with basic functionality landed quickly, I think.

**Comment by vrothberg** [comment]
Great. I assigned @mheon :+1:

**Comment by marty30** [comment]
Is there any progress on this issue? I was looking for a rename feature in a blue-green-deploy. First start a new container with a postfix. Update het mapping in our loadbalancer. Remove the old container. And finally rename the new container to the old one to het back in the original situation without hardly any downtime (only the time that the mapping was updated). For this usecase An alias would indeed also work, which is why I am wondering of there is any progress of alternatives.

**Comment by mheon** [comment]
It's on the agenda, but not an immediate priority - for now, all of our effort is going into the new HTTP API.

**Comment by rhatdan** [comment]
Still in the future.

**Comment by TomSweeneyRedHat** [comment]
FYI, this is now a RFE Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1829029

**Comment by renich** [comment]
@TomSweeneyRedHat well, just tried to check it out and didn't have access to it; even when logged in.

**Comment by TomSweeneyRedHat** [comment]
@renich I'm not sure how the perms work on the Red Hat Bugzilla system.  In short, it basically has the same info as in this issue.  When we have an issue pointing at a BZ or vice versa, I like to set up a pointer the other way so we don't forget to close one out after it's fixed.

**Comment by rhatdan** [comment]
@mheon Want to work on this with an Intern?

**Comment by mheon** [comment]
I think first step is probably to convert all of these various issues requesting renaming, updating, moving into pods, etc into one single `podman container clone` issue.

Beyond that, I need to do a bit of architecture work to figure out how this exactly should look, I don't know for sure how complex this is going to be.

**Comment by JeongHyeon-Kim** [comment]
I was looking for a feature and came here. I hope it will be added soon.

**Comment by rhatdan** [comment]
Best way to get new features is to open PRs :^)

**Comment by rhatdan** [comment]
Sadly coming up on two years on this one

**Comment by rhatdan** [comment]
This one got a little more active this week.

**Comment by rhatdan** [comment]
Still being worked on, not sure if it will make podman 3.0 though?

**Comment by mheon** [comment]
Being worked on in #8955 
Will probably make 3.0 (although in an early and limited form)
